panimg/contrib/oct_converter/readers/e2e.py

```diff
@@ -92,114 +67,89 @@ def __init__(self, filepath):
-                    raw = f.read(20)
-                    image_data = self.image_structure.parse(raw)
-
                     if chunk.ind == 1:  # oct data
-                        all_bits = [
-                            f.read(2)
-                            for i in range(
-                                image_data.height * image_data.width
-                            )
-                        ]
-                        raw_volume = list(
-                            map(self.read_custom_float, all_bits)
-                        )
-                        image = np.array(raw_volume).reshape(
-                            image_data.width, image_data.height
-                        )
-                        image = 256 * pow(image, 1.0 / 2.4)
                         volume_string = f"{chunk.patient_id}_{chunk.study_id}_{chunk.series_id}"
-                        if volume_string in volume_array_dict.keys():
-                            volume_array_dict[volume_string][
-                                int(chunk.slice_id / 2) - 1
-                            ] = image
-                        else:
-                            # try to capture these additional images
-                            if (
-                                volume_string
-                                in volume_array_dict_additional.keys()
-                            ):
-                                volume_array_dict_additional[
-                                    volume_string
-                                ].append(image)
-                            else:
-                                volume_array_dict_additional[volume_string] = [
-                                    image
-                                ]
-                            # print('Failed to save image data for volume {}'.format(volume_string))
+                        volume_dict.append(volume_string)
+                        chunk_stack.append(start)
+
+            # second pass through MDbData chunks:
+            # - extract OCT image data
+            volume_array_dict = {}
+            for volume in set(volume_dict):
+                volume_array_dict[volume] = [0] * len(
+                    [slice for slice in volume_dict if slice == volume]
+                )
+
+            for pos in chunk_stack:
+                f.seek(pos)
+                raw = f.read(60)
+                chunk = self.chunk_structure.parse(raw)
+
+                raw = f.read(20)
+                image_data = self.image_structure.parse(raw)
+
+                all_bits = [
+                    f.read(2)
+                    for i in range(image_data.height * image_data.width)
+                ]
+                raw_volume = list(map(self.read_custom_float, all_bits))
+                image = np.array(raw_volume).reshape(
+                    image_data.width, image_data.height
+                )
+                image = (image * 255).astype(np.uint16)
+                image = 256 * pow(image, 1.0 / 2.4)
```

Comment:
I'm not completely sure if we're doing this correctly. We assumed the 16-bit float value was between 0 and 1 but I just tested it and seemingly the value is between 0 and 2:
```
>>> read_custom_float(0xff, 0xff)
1.9990234375
>>> read_custom_float(0x00, 0x00)
1.0842021724855044e-19
```
Also the casting to uint16 is done too early here. It should be done when the number is in the correct range of values, so after the 1.0/2.4 exponent and 256 factor.

So since we want to scale it to a 16 bit integer we can do it like this:
```
normalized_float = 16bit_float / 2
16bit_scaled_int = (normalized_float * (2^16 - 1)).astype(np.uint16)
```

If we also want to include the exponential (1.0/2.4) we need to integrate that also in the normalizing before scaling. 

-------

I just noticed there is a 16-bit float datatype in both numpy (`np.float16`) and SimpleITK. I think it is even better if we just use that. In the end there will always be some data loss because we simply do not know the exact way the custom 16-bit float datatype in the E2E file is encoded, but this way we are at least using the most similar datatype. So I would only perform the exponential scaling `^(1.0/2.4)` and then cast to `np.float16`. So like this:


```suggestion
                image = (pow(image, 1.0 / 2.4)).astype(np.float16)
```

Reply:
Look liek SimpleITK only support 32-bit and 64-bit floats, but not 16-bit floats. uint16 is the closest then. Will fix the above though!